File: leetcode-cn/0557.0_Reverse_Words_in_a_String_III.py
'''
执行用时：24 ms, 在所有 Python3 提交中击败了98.80% 的用户
内存消耗：15.1 MB, 在所有 Python3 提交中击败了98.57% 的用户
'''
class Solution:
    def reverseWords(self, s: str) -> str:
        return ' '.join(map(lambda word: word[:: -1], s.split()))


'''
执行用时：40 ms, 在所有 Python3 提交中击败了67.86% 的用户
内存消耗：15.5 MB, 在所有 Python3 提交中击败了68.96% 的用户
通过测试用例：29 / 29
'''
class Solution:
    def reverseWords(self, s: str) -> str:
        return ' '.join(''.join(list(reversed(w))) for w in s.split())

'''
执行用时：36 ms, 在所有 Python3 提交中击败了83.39% 的用户
内存消耗：15.5 MB, 在所有 Python3 提交中击败了59.53% 的用户
通过测试用例：29 / 29
'''
class Solution:
    def reverseWords(self, s: str) -> str:
        return ' '.join(w[:: -1] for w in s.split())


'''
执行用时：108 ms, 在所有 Python3 提交中击败了5.67% 的用户
内存消耗：15.4 MB, 在所有 Python3 提交中击败了78.20% 的用户
通过测试用例：29 / 29
'''
class Solution:
    def reverseWords(self, s: str) -> str:
        i, n = 0, len(s)
        ans = ''
        while i < n:
            start = i 
            while i < n and s[i] != ' ':
                i += 1
            left, right = start, i - 1
            while right >= left:
                ans += s[right]
                right -= 1
            while i < n and s[i] == ' ':
                i += 1
                ans += ' '

        return ans 



        

File: cpp/0557.0_Reverse_Words_in_a_String_III.cpp
/*
in place

执行用时：16 ms, 在所有 C++ 提交中击败了18.84% 的用户
内存消耗：9.4 MB, 在所有 C++ 提交中击败了70.49% 的用户
通过测试用例：29 / 29
*/
class Solution {
public:
    string reverseWords(string s) {
        int i = 0, n = s.length();
        while (i < n) {
            int start = i;
            while (i < n && s[i] != ' ') {
                i++;
            }
            int l = start, r = i - 1;
            while (l < r) {
                swap(s[l++], s[r--]);
            }
            while (i < n && s[i] == ' ') {
                i++;
            }
        }
        return s;
    }
};
